File: packages/usf-py/usf/validate.py
"""Validate a skill against the JSON schema + structural rules."""
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

import jsonschema

from .loader import Skill, load
from .parser import REQUIRED_SECTIONS
from .template import find_variables


@dataclass
class ValidationError(Exception):
    skill: str
    errors: list[str]

    def __str__(self) -> str:
        header = f"{self.skill}:"
        return "\n".join([header] + [f"  - {e}" for e in self.errors])


_SCHEMA_CACHE: dict | None = None


def _repo_root() -> Path:
    # packages/usf-py/usf/validate.py -> repo root
    return Path(__file__).resolve().parents[3]


def get_schema() -> dict:
    global _SCHEMA_CACHE
    if _SCHEMA_CACHE is None:
        schema_path = _repo_root() / "schema" / "skill.schema.json"
        _SCHEMA_CACHE = json.loads(schema_path.read_text(encoding="utf-8"))
    return _SCHEMA_CACHE


def validate_skill(skill: Skill) -> list[str]:
    """Return a list of error messages. Empty list = valid."""
    errors: list[str] = []

    # 1. JSON schema on frontmatter
    try:
        jsonschema.validate(skill.frontmatter, get_schema())
    except jsonschema.ValidationError as exc:
        errors.append(f"frontmatter: {exc.message} (at {'/'.join(str(p) for p in exc.path)})")

    # 2. Required sections present
    for s in REQUIRED_SECTIONS:
        if s not in skill.sections:
            errors.append(f"missing required section: # {s}")

    # 3. Inputs declared <-> template variables used
    declared = {i.name for i in skill.inputs_spec}
    used: set[str] = set()
    for body in skill.sections.values():
        used |= find_variables(body)
    undeclared = used - declared
    unused = declared - used
    if undeclared:
        errors.append(f"template uses undeclared input(s): {sorted(undeclared)}")
    if unused:
        errors.append(f"inputs declared but never referenced: {sorted(unused)}")

    return errors


def validate_path(path: str | Path) -> dict[str, list[str]]:
    """Validate every .md in a directory (non-recursive). Returns {name: errors}."""
    p = Path(path)
    results: dict[str, list[str]] = {}
    if p.is_file():
        files = [p]
    else:
        files = sorted(p.glob("*.md"))
    for md in files:
        try:
            skill = load(md)
        except UnicodeDecodeError:
            # Not a text file (e.g. a zipped Claude Skill bundle with .md extension) — skip.
            continue
        except Exception as exc:
            results[str(md)] = [f"parse error: {exc}"]
            continue
        if skill.frontmatter.get("usf") is False:
            continue
        # Only USF-shaped files (have name + description) are validated;
        # everything else is considered non-USF and skipped.
        if "name" not in skill.frontmatter or "description" not in skill.frontmatter:
            continue
        errs = validate_skill(skill)
        if errs:
            results[str(md)] = errs
    return results

File: packages/usf-ts/src/validate.ts
import { promises as fs } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import Ajv2020 from "ajv/dist/2020.js";

import { load, type Skill } from "./loader.js";
import { REQUIRED_SECTIONS } from "./parser.js";
import { findVariables } from "./template.js";

let cachedSchema: Record<string, unknown> | null = null;
let cachedValidator: ReturnType<Ajv2020["compile"]> | null = null;

async function findSchemaPath(): Promise<string> {
  const candidates: string[] = [];
  try {
    const here = path.dirname(fileURLToPath(import.meta.url));
    let dir = here;
    for (let i = 0; i < 8; i++) {
      candidates.push(path.join(dir, "schema", "skill.schema.json"));
      const parent = path.dirname(dir);
      if (parent === dir) break;
      dir = parent;
    }
  } catch {
    // ignore — fall back to CWD walk
  }
  let cwd = process.cwd();
  for (let i = 0; i < 8; i++) {
    candidates.push(path.join(cwd, "schema", "skill.schema.json"));
    const parent = path.dirname(cwd);
    if (parent === cwd) break;
    cwd = parent;
  }
  for (const candidate of candidates) {
    try {
      await fs.access(candidate);
      return candidate;
    } catch {
      // keep looking
    }
  }
  throw new Error(
    `could not locate schema/skill.schema.json (searched from module dir and cwd)`,
  );
}

async function getValidator() {
  if (cachedValidator && cachedSchema) return cachedValidator;
  const schemaPath = await findSchemaPath();
  const text = await fs.readFile(schemaPath, "utf-8");
  cachedSchema = JSON.parse(text);
  const ajv = new Ajv2020({ allErrors: true, strict: false });
  cachedValidator = ajv.compile(cachedSchema as object);
  return cachedValidator;
}

export async function validateSkill(skill: Skill): Promise<string[]> {
  const errors: string[] = [];
  const validator = await getValidator();
  if (!validator(skill.frontmatter)) {
    for (const err of validator.errors ?? []) {
      errors.push(`frontmatter: ${err.message} (at ${err.instancePath || "/"})`);
    }
  }
  for (const s of REQUIRED_SECTIONS) {
    if (!(s in skill.sections)) {
      errors.push(`missing required section: # ${s}`);
    }
  }
  const declared = new Set(skill.inputsSpec.map((i) => i.name));
  const used = new Set<string>();
  for (const body of Object.values(skill.sections)) {
    for (const v of findVariables(body)) used.add(v);
  }
  const undeclared = [...used].filter((v) => !declared.has(v));
  const unused = [...declared].filter((v) => !used.has(v));
  if (undeclared.length) {
    errors.push(`template uses undeclared input(s): ${JSON.stringify(undeclared.sort())}`);
  }
  if (unused.length) {
    errors.push(`inputs declared but never referenced: ${JSON.stringify(unused.sort())}`);
  }
  return errors;
}

export async function validatePath(target: string): Promise<Record<string, string[]>> {
  const results: Record<string, string[]> = {};
  const stat = await fs.stat(target);
  const files = stat.isFile()
    ? [target]
    : (await fs.readdir(target))
        .filter((n) => n.endsWith(".md"))
        .sort()
        .map((n) => path.join(target, n));
  for (const file of files) {
    let skill: Skill;
    try {
      skill = await load(file);
    } catch (exc) {
      const msg = (exc as Error).message;
      if (msg.includes("utf") || msg.includes("encode")) continue;
      results[file] = [`parse error: ${msg}`];
      continue;
    }
    if (skill.frontmatter["usf"] === false) continue;
    if (!skill.frontmatter["name"] || !skill.frontmatter["description"]) continue;
    const errs = await validateSkill(skill);
    if (errs.length) results[file] = errs;
  }
  return results;
}
